[Apps] Allow extra parameters to be passed to action execution

snoe opened the pull request:
For actions, unlike queries, the user will be asked to fill in unmapped
parameters. This change allows the dashboard action execution to accept
incoming parameters without an "id" but with a "target". If that target
exists on the action, it will be accepted.

snoe committed:
[Apps] Allow extra parameters to be passed to action execution

For actions, unlike queries, the user will be asked to fill in unmapped
parameters. This change allows the dashboard action execution to accept
incoming parameters without an "id" but with a "target". If that target
exists on the action, it will be accepted.

kulyk reviewed (approved):
Tried out locally with the FE and it works great 👏

metamben commented on test/metabase/api/dashboard_test.clj:
```diff
@@ -1884,12 +1884,23 @@
+                (is (partial= {:rows-affected 1}
+                              (mt/user-http-request :crowberto :post 200 execute-path
+                                                    {:parameters [{:id "my_id" :type "id" :value 1}
+                                                                  {:target [:variable [:template-tag "name"]]
```
How is this better/more flexlible(?) than `:id "name"`?

metamben reviewed (approved):
Approving to unblock it, but it's not clear to me why things are done the way they are done. I would appreciate talking through it.

snoe commented on src/metabase/actions/execution.clj:
```diff
@@ -36,19 +37,36 @@
+  [incoming-parameters parameter-mappings destination-parameters]
+  (let [destination-parameters-by-target (m/index-by :target destination-parameters)
+        mappings-by-id (m/index-by :parameter_id parameter-mappings)]
+    (mapv (fn [{param-id :id param-target :target :as parameter}]
```
It's hard to know, it certainly seems like they are unnecessary but I'm following how native queries are wired up.

So, let's suppose:
A dashboard for an Order Details page with an `id` parameter (filter). 
A Set Order Tax Rate query action like `update orders set tax_rate = {{tax_rate}} where id = {{order_id}}`
An Order Detail native question to visualize that particular order `select * from orders where id = {{order_id}}`.
The dashboard `id` parameter is tied to the `order_id` parameter on the question card and action button.

This is how things will be saved:
The dashboard will have a `parameters` like `[{:id <dashboard-param-uuid> :slug "id" :name "ID" :type "id"}]`
The dashcards will have a `parameter_mappings` like `[{:parameter_id <dashboard-param-uuid> :target [:variable [:template-tag "order_id"]]}]`
The action and query will have a `parameters` like `[{:id  <card-param-uuid> :target [:variable [:template-tag "order_id"]]}]`

When querying the question, these `parameters` are sent on the request:
`[{:type "id" :value "1" :target [:variable [:template-tag "order_id"]] :id <dashboard-param-uuid>}]`

So you can see:
- The `parameter_mappings` is a mapping from the `<dashboard-param-uuid>` to a `target` 
- There's nothing that maps to the `<card-param-uuid>`
- The `target` on the request parameters is ignored and replaced by the `target` on the `parameter_mappings`.

I felt that introducing the `card-param-uuid` into the request `parameters` would be more confusing than using the `target` that was already being passed through.

The idea in this PR is to be able to provide a parameter that does not have a mapping between dashboard-params and a card's target by simply omitting the `:id <dashboard-param-uuid>`.

metamben replied:
@snoe Thank you for the nice example!

> When querying the question, these `parameters` are sent on the request:
`[{:type "id" :value "1" :target [:variable [:template-tag "order_id"]] :id <dashboard-param-uuid>}]`

This seems to be a mess. What does the URL look like? Either this is a parameter for the question and then it should give a value for a parameter of the question or it's a parameter of the dashboard in which case the BE should map it according to `parameter_mappings`.

Maybe I'm mistaken and this makes perfect sense, but I'm afraid that the pattern we are reusing is not the cleanest and might even go against the goals the Parameters pod had in the last cycle. I wonder if we can do something better we limited effort (especially on FE side). If we can't, I'm OK with repeating an existing pattern.

snoe replied:
> This seems to be a mess. What does the URL look like? Either this is a parameter for the question and then it should give a value for a parameter of the question or it's a parameter of the dashboard in which case the BE should map it according to parameter_mappings.

This is what's passed to 
`POST /api/dashboard/:dashboard-id/dashcard/:dashcard-id/card/:card-id/query` (pre-existing) and 
`POST /api/dashboard/:dashboard-id/dashcard/:dashcard-id/action/:action-id/execute` (new for apps).

I'm not sure what's going on in the front end, but my sense is that there isn't a clear distinction between any of the parameters or mappings and we're likely seeing a bit of "throw it at the wall and see what sticks", or different keys present for querying in other contexts (when building?).

Do you have a suggestion on what would be more in-line with the goals of the Parameters pod? Perhaps, a `target-id` that points to the "destination" parameter's id? ( `card-param-uuid` in this example ).

snoe committed:
Update doc string and validation message

snoe committed:
Merge branch 'master' into data-apps-extra-parameters-on-execution

snoe committed:
Review changes

Remove action-id from the execute endpoint, this will be looked up
through the dashcard.

Revert map-parameters and split parameters into an extra_parameters
value. extra_parameters are combined with the mapped parameters and
passed to execution.

snoe committed:
Fixing tests
